File: csharp/AntFfi/Network.cs
using AntFfi.Native;

namespace AntFfi;

/// <summary>
/// Exception thrown when a network operation fails.
/// </summary>
public class NetworkException : AntFfiException
{
    public NetworkException(string message) : base(message) { }
    public NetworkException(string message, RustCallStatus status) : base(message, status) { }
}

/// <summary>
/// Network configuration for connecting to the Autonomi network.
/// </summary>
public sealed class Network : NativeHandle
{
    /// <summary>
    /// Gets whether this network is configured for local testnet.
    /// </summary>
    public bool IsLocal { get; }

    internal Network(IntPtr handle, bool isLocal) : base(handle)
    {
        IsLocal = isLocal;
    }

    /// <summary>
    /// Creates a new network configuration.
    /// </summary>
    /// <param name="isLocal">If true, connects to local testnet. If false, connects to production network.</param>
    /// <returns>A new Network instance.</returns>
    public static Network Create(bool isLocal)
    {
        var status = RustCallStatus.Create();
        var handle = NativeMethods.NetworkNew(isLocal ? (sbyte)1 : (sbyte)0, ref status);
        if (status.IsError)
            throw new NetworkException("Failed to create network", status);
        return new Network(handle, isLocal);
    }

    /// <summary>
    /// Creates a custom network configuration with specific RPC URL and contract addresses.
    /// </summary>
    /// <param name="rpcUrl">RPC URL for the EVM network (e.g., "http://10.0.2.2:61611").</param>
    /// <param name="paymentTokenAddress">Payment token contract address (hex string).</param>
    /// <param name="dataPaymentsAddress">Data payments contract address (hex string).</param>
    /// <returns>A new Network instance.</returns>
    public static Network CreateCustom(string rpcUrl, string paymentTokenAddress, string dataPaymentsAddress)
    {
        var rpcUrlBuffer = UniFFIHelpers.StringToRustBuffer(rpcUrl);
        var paymentTokenBuffer = UniFFIHelpers.StringToRustBuffer(paymentTokenAddress);
        var dataPaymentsBuffer = UniFFIHelpers.StringToRustBuffer(dataPaymentsAddress);

        var status = RustCallStatus.Create();
        var handle = NativeMethods.NetworkCustom(rpcUrlBuffer, paymentTokenBuffer, dataPaymentsBuffer, ref status);
        if (status.IsError)
            throw new NetworkException("Failed to create custom network", status);
        return new Network(handle, true);
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreeNetwork(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.CloneNetwork(Handle, ref status);
    }
}

File: lua/ant_ffi/network.lua
--[[
  Network configuration for ant_ffi

  This module provides network configuration types:
  - Network: Network configuration for connecting to Autonomi
]]

local ffi = require("ffi")

local M = {}

M._lib = nil
M._helpers = nil
M._errors = nil

function M._init(lib, helpers, errors)
    M._lib = lib
    M._helpers = helpers
    M._errors = errors
end

-- =============================================================================
-- Network
-- =============================================================================

local Network = {}
Network.__index = Network

function Network._wrap(handle)
    local self = setmetatable({}, Network)
    self._disposed = false
    self._handle = ffi.gc(handle, function(h)
        if h ~= nil then
            local status = ffi.new("RustCallStatus")
            M._lib.uniffi_ant_ffi_fn_free_network(h, status)
        end
    end)
    return self
end

--[[
  Create a new Network configuration.
  @param is_local (boolean) - True for local testnet, false for mainnet
  @return Network
]]
function Network.new(is_local)
    assert(M._lib, "network not initialized")
    local is_local_flag = is_local and 1 or 0
    local status = M._errors.new_status()
    local handle = M._lib.uniffi_ant_ffi_fn_constructor_network_new(is_local_flag, status)
    M._errors.check_status(status, "Network.new")
    return Network._wrap(handle)
end

--[[
  Create a custom Network configuration.
  @param rpc_url (string) - RPC endpoint URL
  @param payment_token_address (string) - Payment token contract address
  @param data_payments_address (string) - Data payments contract address
  @return Network
]]
function Network.custom(rpc_url, payment_token_address, data_payments_address)
    assert(M._lib, "network not initialized")
    local rpc_buf = M._helpers.string_to_rustbuffer(rpc_url)
    local token_buf = M._helpers.string_to_rustbuffer(payment_token_address)
    local data_buf = M._helpers.string_to_rustbuffer(data_payments_address)
    local status = M._errors.new_status()
    local handle = M._lib.uniffi_ant_ffi_fn_constructor_network_custom(rpc_buf, token_buf, data_buf, status)
    M._errors.check_status(status, "Network.custom")
    return Network._wrap(handle)
end

function Network:_clone()
    assert(not self._disposed, "Network has been disposed")
    local status = M._errors.new_status()
    local cloned = M._lib.uniffi_ant_ffi_fn_clone_network(self._handle, status)
    M._errors.check_status(status, "Network.clone")
    return cloned
end

function Network:dispose()
    if not self._disposed and self._handle ~= nil then
        ffi.gc(self._handle, nil)
        local status = ffi.new("RustCallStatus")
        M._lib.uniffi_ant_ffi_fn_free_network(self._handle, status)
        self._handle = nil
        self._disposed = true
    end
end

M.Network = Network

return M
